src/charm.py

```diff
@@ -740,6 +721,144 @@ def _gen_wp_config(self):
+            "--admin_password={}".format(admin_password)
+        ]
+
+    def _wp_install(self):
+        """Install WordPress (create WordPress required tables in DB)"""
+        process = self._container().exec(self._wp_install_cmd())
+        process.wait()
+
+    def _init_pebble_layer(self):
+        layer = {
+            "summary": "WordPress layer",
+            "description": "WordPress server",
+            "services": {
+                "wordpress": {
+                    "override": "replace",
+                    "summary": "WordPress server (apache)",
+                    "command": "apache2ctl -D FOREGROUND",
+                }
+            },
+        }
+        if not self._wordpress_service_exists():
+            self._container().add_layer("wordpress", layer, combine=True)
+
+    def _start_server(self):
+        """Start WordPress (apache) server. On leader unit, also make sure WordPress is installed
+
+        Check if the pebble layer has been added, then check the installation status of WordPress,
+        finally start the server. The installation process only run on the leader unit. This
+        operation is idempotence.
+        """
+        self._init_pebble_layer()
+        if self.unit.is_leader():
+            if not self._wp_is_installed():
+                self._wp_install()
+        self._container().start("wordpress")
+
+    @staticmethod
+    def _wp_config_path():
+        return "/var/www/html/wp-config.php"
+
+    def _current_wp_config(self):
+        """Retrieve the current version of wp-config.php from server, return None if not exists"""
+        wp_config_path = self._wp_config_path()
+        container = self._container()
+        if container.exists(wp_config_path):
+            return self._container().pull(wp_config_path).read()
+        return None
+
+    def _push_wp_config(self, wp_config):
+        """Update the content of wp-config.php on server"""
+        self._container().push(self._wp_config_path(), wp_config)
+
+    def _core_reconciliation(self):
+        """Reconciliation process for the WordPress core services, returns True if successful.
+
+        It will fail under the following two circumstances:
+          - Peer relation data not ready
+          - Config doesn't provide valid database information and db relation hasn't
+            been established
+
+        It will check if the current wp-config.php file matches the desired config.
+        If not, update the wp-config.php file.
+
+        It will also check if WordPress is installed (WordPress-related tables exist in db).
+        If not, install WordPress (create WordPress required tables in db).
+
+        If any update is needed, it will stop the apache server first to prevent any requests
+        during the update for security reasons.
+        """
+        if not self._replica_consensus_reached():
+            self.unit.status = WaitingStatus("Waiting for unit consensus")
```

Comment:
Is this the best place to set the status of the charm? Could it, for example, be done by whatever function calls this function based on the return value of this function? Would that make the charm easier to maintain?

Reply:
See the comment below.